### PR TITLE
Sync binding filtering with webapp

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Sync with changes on https://github.com/mattermost/mattermost-webapp/pull/8249

Also piggybacking the changes on https://github.com/mattermost/mattermost-webapp/pull/8782
#### Ticket Link
https://mattermost.atlassian.net/browse/MM-33575
https://mattermost.atlassian.net/browse/MM-33362
https://mattermost.atlassian.net/browse/MM-33436
https://mattermost.atlassian.net/browse/MM-33824

#### Release Note
```release-note
Better binding filtering for Apps Framework
```

### CODE REVIEW

File: app/mm-redux/reducers/entities/apps.ts [hunk 1]
```diff
@@ -11,8 +11,8 @@ import {validateBindings} from '@utils/apps';
 export function bindings(state: AppBinding[] = [], action: GenericAction): AppBinding[] {
     switch (action.type) {
     case AppsTypes.RECEIVED_APP_BINDINGS: {
-        validateBindings(action.data);
-        return action.data || [];
+        action.data = validateBindings(action.data);
+        return action.data;
```
Comment (mickmister) [inline app/mm-redux/reducers/entities/apps.ts:15]: This seems weird that we are editing `action` directly. Some other reducer may access this same action's data at some point right? I don't know if we have control over which reducer receives the action first.

File: app/utils/apps.ts [hunk 2]
```diff
@@ -11,125 +11,193 @@ export function appsEnabled(state: GlobalState) { // eslint-disable-line @typesc
...
-// fillAndTrimBindingsInformation does:
-// - Build the location (e.g. channel_header/binding)
-// - Inherit app calls
-// - Inherit app ids
-// - Trim invalid bindings (do not have an app call or app id at the leaf)
-export function fillAndTrimBindingsInformation(binding?: AppBinding) {
+function cleanBindingRec(binding: AppBinding, topLocation: string, depth: number) {
     if (!binding) {
-        return;
+        return binding;
```
Comment (mickmister) [inline app/utils/apps.ts:20]: Can we put return types on the functions in this file?

**Review by mickmister — APPROVED** [review]
LGTM :+1: I have some non-blocking questions

File: app/mm-redux/reducers/entities/apps.ts (hunk 1 above)
Comment (larkox) [reply app/mm-redux/reducers/entities/apps.ts:15]: validateBindings does update the input value. This is just to be coherent with a similar change in webapp. We can create a new variable here bindings, but the input value would still be updated.

**Review by enahum — APPROVED** [review]
lgtm

**Comment by amyblais** [comment]
@DHaussermann Kind reminder for PR review.

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Tested using steps here https://github.com/mattermost/mattermost-plugin-apps/pull/203#issuecomment-886586970
- All filtering is working as per steps provided
- Functionality for bindings filtering is now consistent with webapp
- Tested on iOS and Android
- No issues found 
LGTM!

@larkox and @amyblais  sorry for the delay on this one.

**Pull request merged**